### PR TITLE
[#5][#9][#10], I adjusted the structure to use the web2py Auth module, a...

### PR DESCRIPTION
**Pull request by OluyemisiAkin**
...dded some validators also. I'm left with creating the unit tests for all the added functionalities

I did this 3days ago, but seems i didn't send the pull request, although i wasn't sure it was necessary

### CODE REVIEW

File: controllers/default.py [hunk 1]
```diff
@@ -71,9 +76,11 @@ def api():
     return Collection(db).process(request,response,rules)
 
 def register():
+    auth.settings.register_next = 'confirm.html'
     response.title = 'iDibo Registration'
-    return dict()
+    return dict(form = auth.register())
 
+@auth.requires_login()
```
Comment (anaoprea) [inline controllers/default.py:68]: is the api() function still needed? you have the same annotation on both, I think we could refactor this

File: models/db.py [hunk 2]
```diff
@@ -1,9 +1,36 @@
-from gluon.tools import Auth
 db = DAL('sqlite://storage.sqlite',pool_size=1,check_reserved=['all'])
+from gluon.tools import Auth, Service, PluginManager
```
Comment (anaoprea) [inline models/db.py]: Please follow PEP8, "Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants": https://www.python.org/dev/peps/pep-0008/#imports

File: models/db.py [hunk 3]
```diff
@@ -1,9 +1,36 @@
-from gluon.tools import Auth
 db = DAL('sqlite://storage.sqlite',pool_size=1,check_reserved=['all'])
```
Comment (anaoprea) [inline models/db.py]: Please follow PEP8, there should be spaces between the arguments of a functions https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements

Pycharm highlights with yellow when something doesn't respect that, you can go on the respective and get tips regarding what is wrong.

File: models/db.py [hunk 4]
```diff
@@ -1,9 +1,36 @@
...
+from gluon.tools import Auth, Service, PluginManager
+
 auth = Auth(db)
 
+## This allows me to still be able to use the web2py auth module while still customising properly the fields
```
Comment (anaoprea) [inline models/db.py]: First, please use PEP8 :) "Each line of a block comment starts with a # and a single space" https://www.python.org/dev/peps/pep-0008/#comments

Second, a personal opinion: I try to keep comments impersonal. I would rephrase it to: "the table allows using the web2py auth module, while customising the fields"

File: models/db.py [hunk 5]
```diff
@@ -1,9 +1,36 @@
...
+                Field('last_name', length=128, default=''),
+                Field('voters_registration_number',),
+                Field('email', length=128, default='', unique=True),
+                Field('password', 'password', length=512,readable=False, label='Password'),
+                Field('registration_key', length=512, writable=False, readable=False, default=''),
+                Field('reset_password_key', length=512, writable=False, readable=False, default=''),
+                Field('registration_id', length=512, writable=False, readable=False, default=''))
+
+#validators
+custom_auth_table = db['auth_user'] # get the custom_auth_table
```
Comment (anaoprea) [inline models/db.py:25]: Please try check PEP8 "Inline comments should be separated by at least two spaces from the statement. They should start with a # and a single space." https://www.python.org/dev/peps/pep-0008/#inline-comments

File: models/db.py [hunk 6]
```diff
@@ -1,9 +1,36 @@
...
+custom_auth_table.first_name.requires = IS_NOT_EMPTY(error_message=auth.messages.is_empty)
+custom_auth_table.last_name.requires = IS_NOT_EMPTY(error_message=auth.messages.is_empty)
+custom_auth_table.password.requires = [IS_STRONG(special=0,upper=0,min=6), CRYPT()]#CRYPT validtor here hadles salting of the pasword
```
Comment (anaoprea) [inline models/db.py]: same as above for inline comments

File: models/db.py [hunk 7]
```diff
@@ -1,9 +1,36 @@
...
+  IS_NOT_EMPTY(),
+  IS_NOT_IN_DB(db, custom_auth_table.email)]
+custom_auth_table.voters_registration_number.requires = [
+    IS_NOT_EMPTY(),
+    IS_NOT_IN_DB(db, custom_auth_table.voters_registration_number)
+]
+
+auth.settings.table_user = custom_auth_table # tell auth to use custom_auth_table
+
+auth.settings.login_next = URL('confirm') #This helps to change the redirected url on user registration or login
```
Comment (anaoprea) [inline models/db.py:41]: I would really not recommend using an inline comment here. You can put this as block comment on the line before and please refer to the comment I made on line 6

File: models/db.py [hunk 8]
```diff
@@ -1,9 +1,36 @@
...
+
 
-auth.settings.extra_fields['auth_user']= [
-    Field('Voters Identification Number', 'string')
-    ]
-auth.define_tables(username=True)
\ No newline at end of file
+auth.define_tables(signature=False)
```
Comment (anaoprea) [inline models/db.py]: From what I see at http://web2py.com/books/default/chapter/29/09/access-control, the "Setting signature=True adds user and date stamping to auth tables, to track modifications.", which is something good for auditing (for example, if a user's account is stolen and the password modified, we could see when that was modified). Why did you disable that?

Comment (anaoprea) [inline models/db.py]: Furthermore, I think the server connection should be encrypted, so using "Auth has an optional secure=True argument, which will force authenticated pages to go over HTTPS." is also useful. If this requires configurations on the web2py server side, we can do it later, I'll add a task for that.

File: controllers/default.py [hunk 9]
```diff
@@ -36,6 +36,11 @@ def user():
         @auth.requires_permission('read','table name',record_id)
     to decorate functions that need access control
     """
+    if request.args[0] == 'register':
+        response.title = 'iDibo Registration'
```
Comment (anaoprea) [inline controllers/default.py]: Since the title of the page is something hardcoded, you could leave it in the static html page where it was before (in views -> default -> register.html -> <title> tag)

File: controllers/default.py [hunk 10]
```diff
@@ -71,9 +76,11 @@ def api():
     return Collection(db).process(request,response,rules)
 
 def register():
+    auth.settings.register_next = 'confirm.html'
     response.title = 'iDibo Registration'
```
Comment (anaoprea) [inline controllers/default.py:64]: You could leave the title of the page hardcoded in the static html, since it's not changing (in views -> default -> register.html -> <title> tag)

File: controllers/default.py [hunk 11]
```diff
@@ -71,9 +76,11 @@ def api():
...
-    return dict()
+    return dict(form = auth.register())
```
Comment (anaoprea) [inline controllers/default.py]: Please follow PEP8 "Don't use spaces around the = sign when used to indicate a keyword argument or a default parameter value." https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements So, the spaces should be line in line 44 above.

File: controllers/default.py (hunk 10 above)
Comment (OluyemisiAkin) [inline controllers/default.py:64]: I tried to do that but the code in layout.html overides it all, so i just left it simply at changing the title in the controller function for the pages as pointed out in the other comment below

File: models/db.py (hunk 8 above)
Comment (OluyemisiAkin) [inline models/db.py]: It seems there are some things in the online book which are missing in the pdf version i downloaded. The downloaded book didn't mention much about the function of signature = True so, i felt it was secure = True that was handling what secure and signature was to do. I also observed earlier before implementing this new method that signature = True displayed the registration key etc as part of the fields in the form. But i've been able to handle that in the readable and writable arguments. I've set it to True now in the corrected work

File: tests/iDibo_tests.py [hunk 12]
```diff
@@ -1,11 +1,24 @@
 from nose.tools import *
```
Comment (anaoprea) [inline tests/iDibo_tests.py]: this is not used anymore, you should have it greyed out in PyCharm; unused code should always be removed

File: tests/iDibo_tests.py [hunk 13]
```diff
@@ -1,11 +1,24 @@
...
-import iDibo
 
-def setup():
-    print "SETUP!"
+import unittest
+from gluon.globals import Request
 
-def teardown():
-    print "TEAR DOWN!"
+db = test_db
```
Comment (anaoprea) [inline tests/iDibo_tests.py:8]: what does this do?

File: tests/iDibo_tests.py [hunk 14]
```diff
@@ -1,11 +1,24 @@
...
+db(db.auth_user.id > 0).delete()
+db.commit()
+
+
+class TestRegister(unittest.TestCase):
+    def setup(self):
+        request = Request()
+
+    def test_register_title(self):
+        self.controller.register()
```
Comment (anaoprea) [inline tests/iDibo_tests.py]: You have created a class "TestRegister", that is extending a "unitest.TestCase"; you are referring to that class through "self" and trying to access the "controller" of that class. However, you have never initialised a instance variable "controller" in "TestRegister" - you have only created "request": this means that the variable "controller" doesn't exist in the scope of your class.

Please have a look at:
- https://docs.python.org/2/tutorial/classes.html
- http://learnpythonthehardway.org/book/ex40.html to 45

Comment (anaoprea) [inline tests/iDibo_tests.py]: By checking http://www.web2py.com/AlterEgo/default/show/260 you can see that the functions in the controller can be accessed directly, so, you should only call "register()"

Comment (anaoprea) [inline tests/iDibo_tests.py]: What does calling "register()" from the controller you mentioned in the "exexfile()" function in line 9 do? You should check what the return type of that function is.

File: tests/iDibo_tests.py [hunk 15]
```diff
@@ -1,11 +1,24 @@
...
+        my_title = self.controller.register.response.title
```
Comment (anaoprea) [inline tests/iDibo_tests.py]: For debugging:
At this level, since there is no controller, you don't have access to any response object, so there is no way to verify the title. You can only verify what you return from the register function, which is dict(form=auth.register()). 
Please read about dictionaries at https://docs.python.org/2/tutorial/datastructures.html
Assuming you don't know what happened, you should check what the response looks like:
    resp = register()
        for key, value in resp.iteritems():
            print "[", key, "]: "
            print "(", value, ")\n"
1. What do you see?
2. Can you access the response title here?

Comment (anaoprea) [inline tests/iDibo_tests.py]: Hint for above: you cannot access the local variables of a function, so either there is a global way of accessing the response of the web2py request or you have to perform assertions on what you returned, which is the form. You could check if it contains "Voters registration number" and the other fields mentioned by using regular expression matching - maybe this should be something for a later phase.

File: tests/iDibo_tests.py [hunk 16]
```diff
@@ -1,11 +1,24 @@
...
+        self.assertEqual(my_title, 'iDibo Registration')
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (anaoprea) [inline tests/iDibo_tests.py]: Please use a test "suite", see https://docs.python.org/2/library/unittest.html#basic-example
This gives you a better control and allows you to run through web2py

File: tests/iDibo_tests.py (hunk 14 above)
Comment (OluyemisiAkin) [inline tests/iDibo_tests.py]: Thats the problem. The functions dont seem to be acesible as register() is still underlined in the code, 
also i found out controller cant be treated like a module so it cant be imported

File: tests/iDibo_tests.py (hunk 16 above)
Comment (OluyemisiAkin) [inline tests/iDibo_tests.py]: maybe d execfile actually worked though.
but i had tried lot of tweaking earlier to access various keys but it keeps saying no value, although it seems it sees the fact that the returned type is actually an sqlform
